0000-error-values-proposal.rst

```diff
@@ -0,0 +1,260 @@
+errors, and defining suitable types that carry the data around until
+it is time to report the errors to the user.
+
+Effect and Interactions
+-----------------------
+
+By turning errors into proper values, tooling authors would be able to
+get rid of their error parsing code and finally be able to concisely
+inspect, render or "decorate" error messages. A few compelling
+applications are listed below:
+
+* A REPL front-end might implement color-coded output, choosing a token's
+  color by its syntactic class (e.g type constructor, data constructor,
+  or identifier), its name (e.g all occurences of ``foldl`` shown in red,
+  occurences of ``concat`` shown in blue), or some other criterion
+  entirely.
+
+* A REPL front-end or IDE tool might allow users the ability to
+  interactively navigate a type in a type error and, for instance, allow
+  the user to interactively expand type synonyms, show kind signatures,
+  etc.
+
+* An IDE tool might ask GHC to defer expensive analyses typically done
+  during error message construction (e.g. `computing valid hole fits
+  <https://gitlab.haskell.org/ghc/ghc/issues/16875#note_210045>`_) and instead
+  query GHC for the analysis result asynchronously (or even only when
+  requested by the user), shrinking the edit/typechecking iteration time.
+
+* An IDE tool might use the suggestions that GHC would embed in error
+  values to present automated refactoring options to the user.
+
+Costs and drawbacks
+-------------------
+
+With the approach described in this proposal, the different error types
+would be scattered accross the different components of the GHC library,
+there would therefore not be a single place to look at to see all the
+errors. They would have to look for all the places in the code where we
+pack some error in a ``GHCError`` wrapper.
+
+Another drawback is that error consumers have to resort to using
+``Typeable`` to implement specific behaviours for any given type, instead
+of pattern matching on constructors of a large sum type of all possible
+errors. Both of those drawbacks come from the dynamic component of the
+design put forward in this proposal, which is essential to avoid the
+import cycles mentionned earlier. We consider the import cycles to be
+a more serious drawback than the dynamic errors, especially given that
+the number of concrete error/warning types should be quite reasonable.
+
+The major cost of implementing this proposal is the sheer amount of
+refactoring that will be necessary to emit error values and move the
+rendering to much later.
+
+Alternatives
+------------
+
+As mentionned in the previous section, we considered a similar approach
+but where there is no existential wrapper and the errors from different
+subsystems are just glued together using sum types. That would have
+required importing all subsystems in one place, introducing import cycles
+that would be very delicate to handle and incur too expensive a price on
```

Comment:
I still don't see why this is so expensive. You will need a hs-boot file that declares `data TcError` and another one for `data RnError`, etc., but that's not really so bad. Or you could use my "abstract data" trick of https://gitlab.haskell.org/ghc/ghc/merge_requests/1957/diffs#576dff4e64fa03683ffc5dcdd58284acdd70bc8e_1175_1209. It just seems unnecessary to foist on all users (IDEs, etc.) of GHC a cost that is only a little inconvenient for us to bear.